Reconnect firehose clients after period of inactivity (timeout for recv)

p1timmy opened the pull request:
Fixes #489

Add a `timeout` parameter to make firehose clients reconnect if no message is received after a given number of seconds instead of waiting forever.

I found 2 ways to set the timeout but not sure which one makes the most sense for most people:

1. Set the parameter in `__init__` to `self._timeout` and pass that value to `client.recv()` (a1fcd3005238b67b34bd7e45f81cd42c5aebe7ef)
2. Use the parameter in `start()` directly as an argument for `client.recv()` (c6b03305fb2d140a67c0c10fb470db0eab547271)

Also allowed the parameter to accept None (the default value for the `timeout` parameter in `client.recv()` which causes it to wait indefinitely for the next message) and gave the following default values for each firehose client type:

- Repos: 30 seconds (because hundreds of ops are being made every second on Bluesky nowadays)
- Labels: 5 minutes (300 seconds)

I plan to squash all these commits once we come to an agreement with which approach to take and get all the other details sorted out.

MarshalX commented:
@p1timmy thank you! i ve fixed some issues around and picked constructor as a way to pass timeout; pls review

p1timmy commented:
Looks good but why is the parameter renamed to `recv_timeout`?

Also just realized the timeout error only forces the client to disconnect, the error has to be handled in code separately for the client to reconnect so will need to update docs to reflect on that.

MarshalX commented:
> why is the parameter renamed to `recv_timeout`?

because "timeout" is too broad name. it could confuse about timeout for the entire client for any activity. for example timeout could be for the websockets framing, etc. and since we use this only in recv, that's why i renamed it like that 

some of the existing timeouts in WebSockets lib that we are using:
- open_timeout
- close_timeout
- ...i guess more

p1timmy commented:
> because "timeout" is too broad name. it could confuse about timeout for the entire client for any activity. for example timeout could be for the websockets framing, etc. and since we use this only in recv, that's why i renamed it like that 
> 
> some of the existing timeouts in WebSockets lib that we are using:
> - open_timeout
> - close_timeout
> - ping_timeout
> - ...i guess more

Oh gotcha, makes a lot of sense now

Also I just squashed all the commits in the branch as you picked the approach to use for making the timeouts work. Other than missing docs on how to reconnect to the server due to timeout errors I think this is ready to merge.

MarshalX commented:
It will reconnect automatically 🧐 that's why I added timeout errors to the errors list of what errors should trigger reconnect mechanism

p1timmy commented:
Oh my bad, that part I didn't pay attention when I reviewed your changes. I just retested my code and the client does reconnect silently after the given timeout, passing `None` restores the old behavior as expected.
